### PR TITLE
[DDP][BE] Fix clang-tidy

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73318 [Easy][c10d] Minor fixes
* **#73299 [DDP][BE] Fix clang-tidy**
* #73237 [DDP][BE] Remove bucket replicas

Differential Revision: [D34422578](https://our.internmc.facebook.com/intern/diff/D34422578)

### CODE REVIEW

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 1]
```diff
@@ -1903,7 +1903,7 @@ namespace {
 // composite key of a tensor's type identifier and its device.
 struct BucketKey {
   BucketKey(c10::ScalarType type, c10::Device device)
-      : type(std::move(type)), device(std::move(device)) {}
```
Comment (awgu) [inline torch/csrc/distributed/c10d/reducer.cpp:1906]: `c10::ScalarType` and `c10::Device` are trivially copyable.

File: torch/csrc/distributed/c10d/logger.cpp [hunk 2]
```diff
@@ -124,11 +124,11 @@ std::vector<std::vector<size_t>> Logger::get_per_bucket_variable_indices() {
...
-std::vector<int> Logger::get_bucket_sizes() {
-  std::vector<int> bucket_sizes;
+std::vector<int64_t> Logger::get_bucket_sizes() {
+  std::vector<int64_t> bucket_sizes;
   for (const auto& bucket : reducer_->buckets_) {
     const auto& variables = bucket.variables;
-    int bucket_size = 0;
+    int64_t bucket_size = 0;
     for (const auto& v : variables) {
       bucket_size += v.numel() * v.element_size();
```
Comment (awgu) [inline torch/csrc/distributed/c10d/logger.cpp:133]: `numel()` returns a `int64_t`, so previously we were (possibly) down-casting to `int`.

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34422578).

**Review by zhaojuanmao — APPROVED** [review]

**Review by mrshenli — COMMENTED** [review]
Thanks @awgu! It's in general good practice to use `init64_t`/`int32_t` instead of int. Could you please also help to update `int` occurrences in `Reducer.hpp/cpp` like below? Thanks!

https://github.com/pytorch/pytorch/blob/7148749e9a7c371379cc12fa4ef6156344f8a8cd/torch/csrc/distributed/c10d/reducer.hpp#L26-L30

**Comment by awgu** [comment]
> Thanks @awgu! It's in general good practice to use `init64_t`/`int32_t` instead of int. Could you please also help to update `int` occurrences in `Reducer.hpp/cpp` like below? Thanks!
> 
> https://github.com/pytorch/pytorch/blob/7148749e9a7c371379cc12fa4ef6156344f8a8cd/torch/csrc/distributed/c10d/reducer.hpp#L26-L30

A pattern of `static_cast<long>(size_t ...)` appears a few times (where `size_t` is unsigned and `long` is signed) in `reducer.cpp`. For example:
https://github.com/pytorch/pytorch/blob/16554bec1b3b5284a9786be6a87dc0ddc3cb2e7c/torch/csrc/distributed/c10d/reducer.cpp#L281-L288
However, `options` sets the dtype of the `local_used_map_` tensor to `int`, so there may be a narrowing conversion (when `sizeof(long) > sizeof(int)`). Are we concerned about these narrowing conversions? If not, then why not explicitly call `static_cast<int>(...)`? This seems like a minor but common issue whenever we have tensors holding `size_t` values since we cannot define tensors with dtype `size_t`.

Generally, what is the thought process behind choosing between 32-bit and 64-bit integer representations in PyTorch? (For example, [here](https://github.com/pytorch/pytorch/blob/16554bec1b3b5284a9786be6a87dc0ddc3cb2e7c/torch/csrc/distributed/c10d/reducer.hpp#L479) uses `long`, while most other places use `int`.)

**Pull request closed without merging**